chia/wallet/cat_wallet/dao_cat_wallet.py

```diff
@@ -0,0 +1,753 @@
+    ) -> DAOCATWallet:
+        self = DAOCATWallet()
+        self.cost_of_single_tx = None
+        self.standard_wallet = wallet
+        self.log = logging.getLogger(__name__)
+
+        limitations_program_hash_hex = bytes32.from_hexstr(limitations_program_hash_hex).hex()  # Normalize the format
+
+        dao_wallet_id = None
+        free_cat_wallet_id = None
+        for id, w in wallet_state_manager.wallets.items():
+            if w.type() == DAOCATWallet.type():
+                assert isinstance(w, DAOCATWallet)
+                if w.get_asset_id() == limitations_program_hash_hex:
+                    self.log.warning("Not creating wallet for already existing DAO CAT wallet")
+                    return w
+            elif w.type() == CATWallet.type():
+                assert isinstance(w, CATWallet)
+                if w.get_asset_id() == limitations_program_hash_hex:
+                    free_cat_wallet_id = w.id()
+        assert free_cat_wallet_id is not None
+        for id, w in wallet_state_manager.wallets.items():
+            if w.type() == WalletType.DAO:
+                self.log.info(f"FOUND DAO WALLET: {w}")
+                self.log.info(f"ALL WALLETS: {wallet_state_manager.wallets}")
+                if w.get_cat_wallet_id() == free_cat_wallet_id:
+                    dao_wallet_id = w.id()
+        assert dao_wallet_id is not None
+        self.wallet_state_manager = wallet_state_manager
+        if name is None:
+            name = CATWallet.default_wallet_name_for_unknown_cat(limitations_program_hash_hex)
+
+        limitations_program_hash = bytes32(hexstr_to_bytes(limitations_program_hash_hex))
+
+        self.dao_cat_info = DAOCATInfo(
+            dao_wallet_id,
+            uint64(free_cat_wallet_id),
+            limitations_program_hash,
+            None,
+            [],
+        )
+        info_as_string = bytes(self.dao_cat_info).hex()
+        self.wallet_info = await wallet_state_manager.user_store.create_wallet(name, WalletType.DAO_CAT, info_as_string)
+
+        self.lineage_store = await CATLineageStore.create(self.wallet_state_manager.db_wrapper, self.get_asset_id())
+        await self.wallet_state_manager.add_new_wallet(self)
+        return self
+
+    async def coin_added(self, coin: Coin, height: uint32, peer: WSChiaConnection) -> None:
+        """Notification from wallet state manager that wallet has been received."""
+        self.log.info(f"DAO CAT wallet has been notified that {coin} was added")
+        # We can't get the inner puzzle for this coin's puzhash because it has the lockup layer.
+        # So look for it's parent coin, and get the inner puzzle for it, which should be the same as
+        # the one contained in the lockup.
+        wallet_node: Any = self.wallet_state_manager.wallet_node
+        parent_coin = (await wallet_node.get_coin_state([coin.parent_coin_info], peer, height))[0]
+        parent_spend = await fetch_coin_spend(height, parent_coin.coin, peer)
```

Comment:
not sure what you meant here - are we not using that utility function already?

Reply:
Hmmmmmmmmmmmmmmmmmmmmm I remember otherwise but github would seem to be gaslighting me into believing this is how it always was and I guess I'm going to let.